### PR TITLE
feat(webhooks): add createFunction/createWebhook authoring primitive (NAN-5885) 1/n

### PR DESCRIPTION
**Pull request by agusayerza**
Introduce `functions` — a trigger-agnostic primitive that separates the trigger (what initiates execution) from the handler (what runs). Webhooks are the first trigger type; `createWebhook()` is sugar for a function with a single implicit `http` trigger. Syncs and actions are unchanged.

- runner-sdk: `createFunction()` / `createWebhook()` + trigger/ingress/debounce types and the `FunctionEvent` shape passed to `exec(nango, event)`.
- types: `ParsedNangoFunction` + `ParsedFunctionTrigger`, additive `functions[]` on the parsed integration (kept separate from ScriptTypeLiteral while additive).
- cli: babel plugin recognizes createFunction/createWebhook and injects `type: 'function'` (distinct const identifier since `function` is reserved); zero-yaml definitions parser builds ParsedNangoFunction; `nango create --webhook` scaffold + template.

Backend deploy acceptance of the `function` type is intentionally out of scope for this slice and tracked as follow-up.

### CODE REVIEW

File: packages/cli/lib/zeroYaml/__snapshots__/compile.unit.cli-test.ts.snap [hunk 1]
```diff
@@ -214,6 +214,6 @@ exports[`edge cases > should catch invalid setMergingStrategy 1`] = `
 
 exports[`edge cases > should catch multiple exports 1`] = `
 "err - ./zero/cases/multipleExports.ts:25 
-  Named export 'test' is not allowed. Only export default and createAction, createSync, createOnEvent are permitted.
+  Named export 'test' is not allowed. Only export default and createAction, createSync, createOnEvent, createFunction, createWebhook are permitted.
```
Comment (agusayerza) [inline packages/cli/lib/zeroYaml/__snapshots__/compile.unit.cli-test.ts.snap:217]: Generated dynamically from allowedExports, I think its fine to leave and leak the upcoming feature here. I don't think exposure will be big

File: packages/cli/templates/webhook.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import { createWebhook } from 'nango';
+
+export default createWebhook({
+    description: '',
+    // Runs synchronously at ingress to verify the request is authentic. No SDK, no I/O.
+    // ingressValidation: async (event) => {
+    //     return true;
+    // },
```
Comment (agusayerza) [inline packages/cli/templates/webhook.ts:8]: Thinking of moving this back to declarative, there are only a handful of validation patterns and the risks of running customer defined code at ingress is real

Comment (marcindobry) [reply packages/cli/templates/webhook.ts:8]: > there are only a handful of validation patterns

Famous last words 😅

File: packages/runner-sdk/lib/scripts.ts [hunk 3]
```diff
@@ -429,3 +430,271 @@ export function createOnEvent<TMetadata extends ZodMetadata = undefined, TCheckp
...
+    const trigger: HttpTrigger = {
+        type: 'http',
+        ...(name !== undefined ? { name } : {}),
+        ...(params.scope !== undefined ? { scope: params.scope } : {}),
+        ...(ingressChallenge !== undefined ? { ingressChallenge } : {}),
+        ...(ingressValidation !== undefined ? { ingressValidation } : {}),
+        ...(debounce !== undefined ? { debounce } : {})
+    };
+    return {
+        type: 'function',
```
Comment (marcindobry) [inline packages/runner-sdk/lib/scripts.ts:702]: Just to highlight that createWebhook is syntactic sugar on top of createFunction, could we actually call createFunction here?

File: packages/cli/templates/webhook.ts (hunk 2 above)
Comment (marcindobry) [reply packages/cli/templates/webhook.ts:8]: Btw, the AI code scanning thing that's my next focus was originally for security / resource usage stuff but it could also be useful here - to check that the non-declarative parts of this are reasonable (if linting is not enough)

**Review by marcindobry — COMMENTED** [review]
Do you have a draft of how createAction/createSync/createOnEvent would be implemented using createFunction? Just want to make sure that we're not going to have to introduce a breaking change to the createFunction interface.

File: packages/cli/templates/webhook.ts (hunk 2 above)
Comment (agusayerza) [reply packages/cli/templates/webhook.ts:8]: Oh yeah, makes sense. That might be good enough to be able to ship it

File: packages/runner-sdk/lib/scripts.ts (hunk 3 above)
Comment (agusayerza) [reply packages/runner-sdk/lib/scripts.ts:702]: Yes. I'm focusing on the webhook use case first, so I wouldn't recommend it yet. But in theory you could and eventually the idea is that is the way of doing it

**Comment by agusayerza** [comment]
I have drafted what `createSync`, `createAction` and `createOnEvent` would look like. `createSync` being the one that is a little bit more complex as it needs to also desugar potential webhook hadnling:

Action and onEvent are pretty straight-forward:
```typescript
export function createOnEvent<...>(
    params: CreateOnEventProps<...>
): CreateFunctionResponse<Record<never, ZodModel>, z.ZodUnknown, z.ZodTypeAny, TMetadata, TCheckpoint> {
    const { event, ...rest } = params;
    const trigger: EventTrigger = { type: 'event', event };
    return {
        type: 'function',
        ...rest,
        triggers: [trigger]
    };
}
```

Action:
```typescript
export function createAction<
    TInput extends z.ZodTypeAny,
    TOutput extends z.ZodTypeAny,
    TMetadata extends ZodMetadata = undefined,
    TCheckpoint extends ZodCheckpoint = undefined
>(params: CreateActionProps<TInput, TOutput, TMetadata, TCheckpoint>): CreateActionResponse<TInput, TOutput, TMetadata, TCheckpoint> {
    const { input, output, exec, endpoint, ...rest } = params;
    const fn = createFunction({
        ...rest,
        triggers: [{ type: 'http' }], 
        input,
        output,
        exec: async (nango, event) => exec(nango, event.payload)
    });
    return { type: 'action', input, output, exec, endpoint, ...rest };
}
```

Syncs:
```typescript
export function createSync<...>(params: CreateSyncProps<...>): CreateFunctionResponse<...> {
    const {
        frequency,
        autoStart,
        syncType: _syncType,           // deprecated, stripped
        trackDeletes: _trackDeletes,   // deprecated, stripped
        onWebhook,
        webhookSubscriptions,
        endpoints: _endpoints,         // deprecated, stripped
        ...rest
    } = params;
    const cronTrigger: CronTrigger = {
        type: 'cron',
        schedule: frequency,
        ...(autoStart !== undefined ? { autoStart } : {})
    };
    const triggers: FunctionTrigger[] = [cronTrigger];
    if (onWebhook !== undefined || webhookSubscriptions !== undefined) {
        // this is the tricky part, haven't settled on what I'll do but was thinking:
        // 1. Deprecate the fields once this gets released (and provide a migration strategy on the cli)
        // 2. Create a new http trigger on this sync that will handle the webhook, but piping the onWebhook function might be tricky
        // 3. calling createWebhook either here or when parsing the functions and synthesise the webhook 
        // Those options are not in order of preference, and this probably needs a bigger discussion with everyone. Interface will look the same regardless
    }
    return { type: 'function', ...rest, triggers };
}
```

cc @marcindobry

File: packages/cli/lib/services/ensure.service.ts [hunk 4]
```diff
@@ -38,12 +38,14 @@ export class Ensure {
         }
     }
 
-    public async functionType(sync: boolean, action: boolean, onEvent: boolean): Promise<FunctionType> {
```
Comment (TBonnin) [inline packages/cli/lib/services/ensure.service.ts:41]: unrelated to this PR but ordering of params is meaningful and error prone. Making it an object would be clearer and less likely to use it wrong

File: packages/runner-sdk/lib/scripts.ts [hunk 5]
```diff
@@ -429,3 +430,271 @@ export function createOnEvent<TMetadata extends ZodMetadata = undefined, TCheckp
...
+    /** Maps to the webhook URL path segment. Defaults to the file basename. */
+    name?: string;
+    /** `connection` opts into connection-level URLs (`/:webhookName/:targetToken`). */
+    scope?: 'integration' | 'connection';
+    ingressChallenge?: IngressChallenge;
+    ingressValidation?: IngressValidation;
+    debounce?: WebhookDebounceConfig;
+}
+export interface CronTrigger {
+    type: 'cron';
```
Comment (TBonnin) [inline packages/runner-sdk/lib/scripts.ts]: let's not call it `cron` since it doesn't follow cron syntax or semantic

File: packages/runner-sdk/lib/scripts.ts [hunk 6]
```diff
@@ -429,3 +430,271 @@ export function createOnEvent<TMetadata extends ZodMetadata = undefined, TCheckp
...
+ *     }
+ * });
+ * ```
+ */
+export function createWebhook<
+    TModels extends Record<string, ZodModel> = Record<never, ZodModel>,
+    TMetadata extends ZodMetadata = undefined,
+    TCheckpoint extends ZodCheckpoint = undefined
+>(params: CreateWebhookProps<TModels, TMetadata, TCheckpoint>): CreateFunctionResponse<TModels, z.ZodTypeAny, z.ZodTypeAny, TMetadata, TCheckpoint> {
+    const { name, ingressChallenge, ingressValidation, debounce, ...rest } = params;
```
Comment (TBonnin) [inline packages/runner-sdk/lib/scripts.ts]: should also desctructure `scope` otherwise it will stay in rest and be applied at top level in addition to behing inside trigger

File: packages/cli/lib/zeroYaml/compile.ts [hunk 7]
```diff
@@ -489,6 +503,36 @@ function nangoPlugin({ entryPoint }: { entryPoint: string }) {
     return {
         bag,
         plugin: ({ types: t }: { types: typeof babel.types }): babel.PluginObj<any> => {
+            // Properties that `createWebhook()` lifts into its implicit `http` trigger at runtime.
+            // Keep in sync with `createWebhook()` in runner-sdk/lib/scripts.ts.
+            const webhookTriggerFields = new Set(['name', 'scope', 'ingress', 'ingressChallenge', 'ingressValidation', 'debounce']);
```
Comment (TBonnin) [inline packages/cli/lib/zeroYaml/compile.ts]: is 'ingress` being used anywhere?

File: packages/runner-sdk/lib/scripts.ts [hunk 8]
```diff
@@ -429,3 +430,271 @@ export function createOnEvent<TMetadata extends ZodMetadata = undefined, TCheckp
...
+export interface CreateFunctionProps<
+    TModels extends Record<string, ZodModel>,
+    TInput extends z.ZodTypeAny = z.ZodUnknown,
+    TOutput extends z.ZodTypeAny = z.ZodTypeAny,
+    TMetadata extends ZodMetadata = undefined,
+    TCheckpoint extends ZodCheckpoint = undefined
+> {
+    /**
+     * The version of the function. Use it to track changes inside Nango's UI.
+     * @default '0.0.1'
```
Comment (TBonnin) [inline packages/runner-sdk/lib/scripts.ts]: is it true version defaults to 0.0.1?

File: packages/runner-sdk/lib/scripts.ts [hunk 9]
```diff
@@ -429,3 +430,271 @@ export function createOnEvent<TMetadata extends ZodMetadata = undefined, TCheckp
...
+// execution) from the handler (what runs). Webhooks are the first trigger type.
+// `createWebhook()` is syntactic sugar for a function with a single implicit `http` trigger.
+// Syncs and actions will eventually be merged into functions.
+
+/**
+ * The kind of trigger that initiated a function run.
+ * - `http`: an incoming http call or webhook request
+ * - `cron`: a periodic schedule
+ * - `event`: an internal Nango event
+ * - `manual`: started via `nango.triggerFunction()`, the API, or the CLI
```
Comment (TBonnin) [inline packages/runner-sdk/lib/scripts.ts]: `manual` implies it is triggered by a human but it could be triggered by any other automated system. Should manual even be modeled? is it equivalent to no have a trigger?

**Review by TBonnin — APPROVED** [review]

File: packages/runner-sdk/lib/scripts.ts [hunk 10]
```diff
@@ -429,3 +430,273 @@ export function createOnEvent<TMetadata extends ZodMetadata = undefined, TCheckp
...
+// execution) from the handler (what runs). Webhooks are the first trigger type.
+// `createWebhook()` is syntactic sugar for a function with a single implicit `http` trigger.
+// Syncs and actions will eventually be merged into functions.
+
+/**
+ * The kind of trigger that initiated a function run.
+ * - `http`: an incoming http call or webhook request
+ * - `schedule`: a periodic schedule
+ * - `event`: an internal Nango event
+ * - `api`: invoked directly via `nango.triggerFunction()`, the REST API, the CLI, or the UI
```
Comment (TBonnin) [inline packages/runner-sdk/lib/scripts.ts]: is FunctionTrigger type exposed to customers? if if it purely internal 'api' might be ok but if it is exposed to customers it might ber ambiguous what 'api' means (direct api vs indirect via CLI UI, node client)

Comment (agusayerza) [reply packages/runner-sdk/lib/scripts.ts]: Dropping it. Went back to the design and we had settled on it being an implicit trigger as every function will support being called by the API, CLI or UI
